src/dart/codegen.rs

```diff
@@ -1,129 +1,215 @@
 use dsl::ast::*;
-use dart::lex::{Lexer, Token};
+use dart::ast;
 use syntax::symbol::Symbol;
+use node::Node;
 
-pub struct Codegen {
-    tokens: Vec<Token>,
-}
+pub struct Codegen {}
 
 impl Codegen {
     pub fn new() -> Self {
-        Codegen { tokens: vec![] }
-    }
-
-    fn write_str(&mut self, string: &str) {
-        let fm = ::codemap().new_filemap(String::new(), string.to_string());
-        self.tokens.extend(
-            Lexer::new(::mk_sp(fm.start_pos, fm.end_pos))
-                .tokenize()
-                .unwrap()
-                .into_iter()
-                .map(|(_, t)| t),
-        );
-    }
-
-    fn write_ident(&mut self, ident: &Symbol) {
-        self.tokens.push(Token::Identifier(*ident));
+        Codegen {}
     }
 
-    pub fn codegen_items(mut self, items: &[Item]) -> Vec<Token> {
-        for item in items {
-            self.codegen_item(item);
-        }
-        self.tokens
+    pub fn codegen_items(&mut self, items: &[Item]) -> Vec<Node<ast::Item>> {
+        //search for name, imports/exports, directives
+        items
+            .iter()
+            .flat_map(|item| self.codegen_item(item))
+            .collect()
     }
 
-    fn codegen_item(&mut self, item: &Item) {
+    fn codegen_item(&mut self, item: &Item) -> Vec<Node<ast::Item>> {
         match *item {
-            Item::ComponentDef(ref name, ref fields, ref instance) => {
-                self.write_str("class ");
-                self.write_ident(name);
-                self.write_str(" extends StatelessWidget {\n");
+            Item::ComponentDef(name, ref fields, ref instance) => {
+                let params;
+                let body;
+                let superclass;
+                let mut class_members = vec![];
+                superclass = Some(ast::Type::simple_path("StatelessWidget"));
                 if !fields.is_empty() {
-                    self.write_str("const ");
-                    self.write_ident(name);
-                    self.write_str("({ Key key");
+                    let mut normal = vec![];
+                    let mut default = vec![];
+                    normal.push(ast::ArgDef::build_param(
+                        ast::Type::simple_path("Key"),
+                        false,
+                        "key",
+                    ));
+
                     for field in fields {
-                        self.write_str(", this.");
-                        self.write_ident(&field.name);
+                        let mut var_ty = Node::new(ast::Type::Infer);
+                        if let Some(ref ty) = field.ty {
+                            var_ty = self.codegen_type(ty);
+                        }
+                        let arg = ast::ArgDef::build_param(var_ty, true, &field.name.as_str());
```

Comment:
This should manually create the `ArgDef` instead of going through `.as_str()` (`field.name` is already a `Symbol`).